scripts/nlp_language_modeling/build_knn_map_index.py

```diff
@@ -201,22 +292,82 @@ def get_emb():
     group.add_argument('--vocab-file', type=str, default=None, help='Path to the vocab file')
     group.add_argument('--merge-file', type=str, default=None, help='Path to the BPE merge file (if necessary).')
     group.add_argument('--delimiter', type=str, default=None, help='delimiter used for tabular tokenizer')
+    group.add_argument(
+        '--stage',
+        type=int,
+        default=None,
+        help='used for building the large knn index in multiple stages',
+        choices=[1, 2],
+    )
+    group.add_argument('--workers', type=int, default=None, help='number of workers to run tokenizer')
+    group.add_argument(
+        '--nprobe',
+        type=int,
+        default=10,
+        help='number of probes, higher number of probes renders better results but runs slower',
+    )
+    group.add_argument(
+        '--shard_index_input',
+        type=str,
+        default=None,
+        help='the knn sharding index files, which are created at stage 1',
+    )
 
     args = parser.parse_args()
+
+    has_gpu = torch.cuda.is_available() and hasattr(faiss, "index_gpu_to_cpu")
+
+    if not hasattr(faiss, "index_gpu_to_cpu"):
+        logging.warning(
+            "faiss doesn't support gpu index. Please check https://github.com/facebookresearch/faiss/blob/main/INSTALL.md"
+        )
+
+    if args.stage == 2:
+        # combine shard index files into one
+        input_file = pathlib.Path(args.shard_index_input)
+        path = input_file.parent
+        fname = input_file.name
+        all_files = [str(i) for i in pathlib.Path(path).glob(fname + '*')]
+        merge_knn_files(all_files, args.output_file)
+        f = KNNIndex(args.output_file)
+        logging.info(f'Write to {args.output_file},  Size of Index : {f.len}')
+        logging.info(f'Index neighbors: {f.K}')
+        logging.info(f'Index chunk start id: {f.chunk_start_id}')
+        logging.info(f'Index chunk end id: {f.chunk_end_id}')
+        sys.exit(0)
```

Comment:
Can you add a few more high level comments in the code below explaining the purpose of code blocks?

Reply:
I do have a high level comment about the script at the beginning of the file with some examples of how to use it.